In [1]:
import numpy as np
import pandas as pd
import statistics as st

In [2]:
students = pd.read_csv(r'C:\Users\USER\Downloads\StudentsPerformance.csv')

In [3]:
students

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
0                      none           0             17             10  
1                      none           8             24             23  
2                      none          18             32             28  
3                      none          11             38             32  
4                      none          22             39             33  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[1000 rows x 8 columns]

In [4]:
students.shape

(1000, 8)

In [5]:
students.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group C            some high school  free/reduced   
1  female        group B                 high school  free/reduced   
2  female        group B            some high school  free/reduced   
3  female        group B                some college      standard   
4  female        group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none           0             17             10  
1                    none           8             24             23  
2                    none          18             32             28  
3                    none          11             38             32  
4                    none          22             39             33

In [6]:
students.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
gender                         1000 non-null object
race/ethnicity                 1000 non-null object
parental level of education    1000 non-null object
lunch                          1000 non-null object
test preparation course        1000 non-null object
math score                     1000 non-null int64
reading score                  1000 non-null int64
writing score                  1000 non-null int64
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [7]:
students.describe()

math score  reading score  writing score
count  1000.000000    1000.000000    1000.000000
mean     66.001000      69.195000      68.119000
std      15.569567      14.706436      15.226381
min       0.000000      17.000000      10.000000
25%      57.000000      59.000000      58.000000
50%      66.000000      70.000000      69.000000
75%      77.000000      79.250000      79.000000
max     100.000000     100.000000     100.000000

#QUESTION 1. To find out how many males and females participated in the test:

In [9]:
students['gender'].value_counts()

female    518
male      482
Name: gender, dtype: int64

Inference: There were 518 female students and 482 male students who participated in the test.

In [10]:
students.select_dtypes('object').nunique()

gender                         2
race/ethnicity                 5
parental level of education    6
lunch                          2
test preparation course        2
dtype: int64

QUESTION 2 To analyze the students' parental level of education:

In [11]:
students['parental level of education'].value_counts()

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

Inference: The majority of the students' parents have completed some college education.

QUESTION 3 .To analyse the  average for math, reading and writing based on test preparation couse and gender


In [12]:
students[['test preparation course','gender','math score','reading score','writing score']].groupby(['test preparation course','gender']).agg('mean')

math score  reading score  writing score
test preparation course gender                                          
completed               female   67.108696      77.375000      78.793478
                        male     72.988506      70.793103      70.339080
none                    female   61.320359      69.955090      68.982036
                        male     66.467532      62.581169      59.551948

Inference:

Female students who completed the test preparation course had the highest average scores for reading, and writing, and those who didnot completed scored the highest in reading and writing

male students who completed the test preparation course had the highest average math score, and males who did not complete the test preparation course score the highest in math.

#QUESTION 4. To analyze the scoring variation for math, reading, and writing based on Gender and Test preparation course:

In [15]:
students[['test preparation course','gender','math score','reading score','writing score']].groupby(['test preparation course','gender']).agg('var')

math score  reading score  writing score
test preparation course gender                                          
completed               female  208.173913     161.175546     149.836897
                        male    197.098133     185.297987     166.098233
none                    female  272.602767     214.241220     225.495172
                        male    205.513600     184.843553     181.394687

Inference: There is more variation in the scores of female students who did not complete the test preparation course compared to those who did. There is relatively less variation in the scores of male students.

#QUESTION 5. To give bonus points to the top 25% of students based on their math score

In [14]:
# calculate the 25th percentaile of math score
math_percentile = students['math score'].quantile(0.75)

top_students = students[students['math score']>= math_percentile]

# assign bonus point to the 25% of students based on their math score
top_students['bonus points'] = 5

print(top_students.head())

     gender race/ethnicity parental level of education         lunch  \
414  female        group A                 high school  free/reduced   
415  female        group B             master's degree  free/reduced   
416  female        group B           bachelor's degree  free/reduced   
417  female        group B             master's degree      standard   
418  female        group B                 high school      standard   

    test preparation course  math score  reading score  writing score  \
414               completed          77             88             85   
415               completed          77             97             94   
416                    none          77             85             87   
417                    none          77             90             84   
418               completed          77             82             89   

     bonus points  
414             5  
415             5  
416             5  
417             5  
418             5  


C:\Users\USER\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


The top 25% of students based on their math scorehave been assigned 5 bonus points